web/src/stream_topic_history.ts

```diff
@@ -46,21 +67,49 @@ export function all_topics_in_cache(sub: StreamSubscription): boolean {
         return true;
     }
 
-    // Now, we can just compare the first cached message to the first
-    // message ID in the stream; if it's older, we're good, otherwise,
-    // we might be missing the oldest topics in this stream in our
-    // cache.
     const first_cached_message = all_messages_data.first();
-    return first_cached_message!.id <= sub.first_message_id;
+    if (sub.first_message_id < first_cached_message!.id) {
+        // Missing the oldest topics in this stream in our cache.
+        return false;
+    }
+
+    // At this stage, number of topics displayed in the topic list
+    // widget is at max `topic_list_data.max_topics` and
+    // sub.first_message_id >= first_cached_message!.id.
+    //
+    // There's a possibility of a few topics missing for messages
+    // which were sent when the user wasn't subscribed.
+    // Fetch stream history to confirm if all topics are already
+    // displayed otherwise rebuild with updated data.
+    const url = "/json/users/me/" + sub.stream_id + "/topics";
+    channel.get({
```

Comment:
I feel like maybe this should be just a `server_topic_history_util.get_server_history` call, with `update_topic_list_for_stream` as a registered callback?

To avoid import cycles, we'd want to move this function to `topic_list.ts` rather than having ti be here, but that might be an improvement in code organization anyway?

There is a bug in that approach, which is that we should be appending the callback to a callback list inside the `stream_topic_history.is_request_pending_for` care, but that bug is existing and could be readily fixed by maintaining an array of callbacks for a given channel.